### ISSUE TITLE
pay should not fail if success is still possible

### ISSUE THREAD

**Issue by rustyrussell** [ElementsProject/lightning#2629]
Shadow on the lightning-makers telegram points out:

```
try:
  clightning.pay(bolt11)
  charge_balance(user)
except RpcError as e:
  handle_rpc_err(e)
```

Payment can still be pending with the Timeout error though!  Especially in the case of a hold invoice.  This is terrible UX!

**Comment by cdecker** [comment]
Agreed, hodl invoices are a terrible idea :wink:

**Comment by ZmnSCPxj** [comment]
The issue is that payment can still succeed even after the daemon is shut down.

Possibly what we need is some kind of system that is more similar to the inverse of the `invoice` interface, with such things as `waitanypay` or `waitpay`...

And proper wallets should *initiate* payment with `pay`, which returns immediately, then monitors events via `waitpay` and `waitanypay`.

Then timeout can be implemented by wallet software instead.

**Comment by fiatjaf** [comment]
That code is wrong no matter what. The balance should be charged before the payment attempt is even started.

But besides that there's already `sendpay` and `waitsendpay`, right?

**Comment by ZmnSCPxj** [comment]
See, the problem is that there are really three possible results:

1.  Definite success.
2.  Definite failure.
3.  We do not know, something is stalled along the route.

Client code should really differentiate between result 3 from the other results.
The shown client code assumes only result 1 and result 2.

Our RPC can only signal "definite success" and "definite failure".
The only alternative is for RPC to stall indefinitely, but that is another UX problem (especially if wallets are naively implemented with single-threaded blocking RPC for a GUI; worse if there is only one JSON-RPC connection the wallet uses).

This is the reason why long, long ago I proposed having a separate *script* for `pay`, which would continuously print out status, rather than returning some result that can be misinterpreted as above.

Maybe the recent changes to RPC for the million-channels-project optimizations may allow us to print partial results for the `pay` command continuously, which savvy wallets can use to monitor state.

**Comment by ZmnSCPxj** [comment]
Oh no.

It looks like the pay plugin regressed the previous behavior.

Previously, `pay` would return `PAY_IN_PROGRESS` (error 200) if it timed out while a `sendpay` was still active.

Now we seem to never return `PAY_IN_PROGRESS` from `pay`.

First part is to return `PAY_IN_PROGRESS` if a `waitsendpay` is in progress.

**Comment by fiatjaf** [comment]
> I decided to fold "we do not know" to the "failure" case.

This is wrong. If anything, "we do not know" should be treated as a success by clients.

**Comment by ZmnSCPxj** [comment]
>This is wrong. If anything, "we do not know" should be treated as a success by clients.

But it *could* stall and *then* fail, and the client has to know that it ultimately failed --- the payment did not go through after stalling.

Either way, it should *not* be treated as *either* success or failure by clients, which is why `PAY_IN_PROGRESS` exists at all.
Mistreating "we do not know" as success is just as bad as mistreating it as failure.

The Real Problem is that our RPC cannot provide intermediate results while the attempt is ongoing, which is why `pay` was originally designed in such a bad UX way --- we needed to be able to report to users "we tried paths X and Y and Z and currently we are stuck with path Z for the past minute already".
